java-squid/src/main/java/org/sonar/java/resolve/JavaType.java

```diff
@@ -384,5 +389,27 @@ public JavaType substitution(TypeVariableJavaType typeVariableType) {
       }
       return Lists.newArrayList();
     }
+
+    @Override
+    public boolean isSubtypeOf(Type superType) {
+      if (superType instanceof JavaType.ParametrizedTypeJavaType && erasure().isSubtypeOf(superType.erasure())) {
+        return checkWildCards((JavaType.ParametrizedTypeJavaType) superType);
+      }
+      return super.isSubtypeOf(superType);
+    }
+
+    private boolean checkWildCards(ParametrizedTypeJavaType superType) {
+      Map<String, JavaType> itsTypes = new HashMap<>();
+      for (Map.Entry<TypeVariableJavaType, JavaType> entry : superType.typeSubstitution.substitutionEntries()) {
+        itsTypes.put(entry.getKey().name(), entry.getValue());
+      }
+      for (Map.Entry<TypeVariableJavaType, JavaType> entry : typeSubstitution.substitutionEntries()) {
+        JavaType itsType = itsTypes.get(entry.getKey().name());
+        if (Symbols.anyType != itsType) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)